metrics/density/fast_footprint.sh

```diff
@@ -0,0 +1,556 @@
+# - which container workload we run
+# - which container runtime we run with
+# - when do we terminate the test (cutoff points)
+#
+# There are a number of things we may wish to add to this script later:
+# - sanity check that the correct number of runtime components (qemu, shim etc.)
+#  are running at all times
+# - some post-processing scripts to generate stats and graphs
+#
+# The script gathers information about both user and kernel space consumption
+# Output is into a .json file, named using some of the config component names
+# (such as footprint-busybox.json)
+
+# Pull in some common, useful, items
+SCRIPT_PATH=$(dirname "$(readlink -f "$0")")
+source "${SCRIPT_PATH}/../lib/common.bash"
+
+# Note that all vars that can be set from outside the script (that is,
+# passed in the ENV), use the ':-' setting to allow being over-ridden
+
+# Default sleep for 3s to let containers come up and finish their
+# initialisation before we take the measures. Some of the larger
+# containers can take a number of seconds to get running.
+PAYLOAD_SLEEP="${PAYLOAD_SLEEP:-10}"
+
+# How long, in seconds, do we wait for KSM to 'settle down', before we
+# timeout and just continue anyway.
+KSM_WAIT_TIME="${KSM_WAIT_TIME:-300}"
+
+# How long, in seconds, do we poll for docker to complete launching all the
+# containers?
+DOCKER_POLL_TIMEOUT="${DOCKER_POLL_TIMEOUT:-300}"
+
+# How many containers do we launch in parallel before taking the PAYLOAD_SLEEP
+# nap
+PARALLELISM="${PARALLELISM:-10}"
+
+### The default config - run a small busybox image
+# Define what we will be running (app under test)
+#  Default is we run busybox, as a 'small' workload
+PAYLOAD="${PAYLOAD:-busybox}"
+PAYLOAD_ARGS="${PAYLOAD_ARGS:-tail -f /dev/null}"
+PAYLOAD_RUNTIME_ARGS="${PAYLOAD_RUNTIME_ARGS:- -m 2G}"
+
+
+#########################
+### Below are a couple of other examples of workload configs:
+#  mysql is a more medium sized workload
+#PAYLOAD="${PAYLOAD:-mysql}"
+# Disable the aio use, or you can only run ~24 containers under runc, as you run out
+# of handles in the kernel. Disable log-bin as it hits a file resize fail error.
+#PAYLOAD_ARGS="${PAYLOAD_ARGS:- --innodb_use_native_aio=0 --disable-log-bin}"
+#PAYLOAD_RUNTIME_ARGS="${PAYLOAD_RUNTIME_ARGS:- -m 4G -e MYSQL_ALLOW_EMPTY_PASSWORD=1}"
+#
+#  elasticsearch is a large workload
+#PAYLOAD="${PAYLOAD:-elasticsearch}"
+#PAYLOAD_ARGS="${PAYLOAD_ARGS:-}"
+#PAYLOAD_RUNTIME_ARGS="${PAYLOAD_RUNTIME_ARGS:- -m 8G}"
+#########################
+
+###
+# which RUNTIME we use is picked up from the env in
+# common.bash. You can over-ride by setting RUNTIME in your env
+
+###
+# Define the cutoff checks for when we stop running the test
+  # Run up to this many containers
+NUM_CONTAINERS="${NUM_CONTAINERS:-200}"
+  # Run until we have consumed this much memory (from MemFree)
+MAX_MEMORY_CONSUMED="${MAX_MEMORY_CONSUMED:-256*1024*1024*1024}"
+  # Run until we have this much MemFree left
+MIN_MEMORY_FREE="${MIN_MEMORY_FREE:-2*1024*1024*1024}"
```

Comment:
Like `PAYLOAD_RUNTIME_ARGS` above, it would be good to check whether these values are realistic for the system being tested, exiting before (potentially very-long running) tests are started. I appreciate it's up to users to set these values, but it doesn't hurt having a safety net to avoid the potential headaches of trying to debug weird behaviour caused by `ENOMEM` ;)